### PR TITLE
Adding default XLA/GPU env vars for all JAX-based containers

### PR DESCRIPTION
**Pull request by terrykong**
Addresses #105 and #109 

*  `--xla_gpu_enable_latency_hiding_scheduler=true`: Allows XLA:GPU to move communication collectives to increase overlap with compute kernels.
* `--xla_gpu_enable_async_all_gather=true`: Allows XLA:GPU to run All Gather NCCL kernels on a separate CUDA stream to allow overlap with compute kernels.
* `--xla_gpu_enable_async_reduce_scatter=true`: Allows XLA:GPU to run Reduce Scatter NCCL kernels on a separate CUDA stream to allow overlap with compute kernels.
* `--xla_gpu_enable_triton_gemm=false`: Disallows Trition GeMM kernels; uses CUBLAS GeMM kernels instead. CUBLAS kernels are currently better tuned for GPUs and thus provide better performance.
* `CUDA_DEVICE_MAX_CONNECTIONS=1`: Use a single queue for GPU work, lowers latency of each stream operation. OK since XLA already orders launches.
* `NCCL_IB_SL=1`: defines the InfiniBand Service Level ([1](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-ib-sl))

@nluehr @ashors1

### CODE REVIEW

File: .github/container/Dockerfile.jax [hunk 1]
```diff
@@ -47,6 +47,11 @@ ARG SRC_PATH_JAX
 ARG SRC_PATH_XLA
 ARG BUILD_DATE
 ENV BUILD_DATE=${BUILD_DATE}
+ENV XLA_FLAGS="--xla_gpu_enable_latency_hiding_scheduler=true --xla_gpu_enable_async_all_gather=true --xla_gpu_enable_async_reduce_scatter=true --xla_gpu_enable_triton_gemm=false"
```
Comment (nluehr) [inline .github/container/Dockerfile.jax:50]: @abhinavgoel95 should we also include `xla_gpu_enable_pipelined_all_reduce` and `xla_gpu_enable_pipelined_all_gather` to this list?

Comment (abhinavgoel95) [reply .github/container/Dockerfile.jax:50]: No, we see perf issues with those flags as of today. Hoping they get fixed by early next week.

**Comment by yhtang** [comment]
Could someone please edit the PR description to document the rationale behind each flag?

**Comment by terrykong** [comment]
> Could someone please edit the PR description to document the rationale behind each flag?

@nluehr @abhinavgoel95 ?

**Comment by yhtang** [comment]
- `NCCL_PROTO=LL128`:
  [doc](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-proto) says:
  > The default is LL,LL128,Simple on platforms which support LL128, LL,Simple otherwise.
  
  Related issue: https://github.com/NVIDIA/nccl/issues/281

- NCCL_AVOID_RECORD_STREAMS=1: this is undocumented in https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html.

- NCCL_IB_SL=1
  [doc](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/env.html#nccl-ib-sl) says:
  > Defines the InfiniBand Service Level.
  > For more information, see the InfiniBand specification Volume 1 (https://www.infinibandta.org/ibta-specifications-download) or vendor documentation.
  
  and the [link](https://www.infinibandta.org/ibta-specifications-download) in the quoted text above is **dead**.

**Comment by nluehr** [comment]
@abhinavgoel95 How much speedup do we see from setting `NCCL_PROTO=LL128` in practice?
The combiner threshold may come into play here as for larger collectives, I think LL128 could reduce effective bandwidth by ~6%.

**Comment by nluehr** [comment]
From my reading of the NCCL code, the only safe way to enable LL128 is to not specify NCCL_PROTO at all.
LL128 is valid only on Volta, Ampere, and Hopper with NVLink. PCIE can cause packet re-ordering that results in silent corruption.
Since we expect users will run our models on a variety of clusters, I think we shouldn't define NCCL_PROTO either in the JAX dockerfile nor in the paxml run scripts. If removing this results in worse performance, we should open a bug with the NCCL team.

**Comment by ashors1** [comment]
We're also sometimes interested in setting `XLA_PYTHON_CLIENT_MEM_FRACTION=0.9` to increase the amount of memory allocated to XLA. Is this something we'd want to consider adding to the base container as well?

**Comment by nluehr** [comment]
Whether we want to set XLA_PYTHON_CLIENT_MEM_FRACTION depends on the workload (particularly whether any other GPU libraries are being used outside of XLA).

My opinion is that this would be OK to set in the paxml container, but for the JAX container in general we shouldn't set it because there we expect users to extend it in varied workflows.

File: .github/container/Dockerfile.jax [hunk 2]
```diff
@@ -47,6 +47,11 @@ ARG SRC_PATH_JAX
...
+ENV CUDA_DEVICE_MAX_CONNECTIONS=1
+ENV NCCL_AVOID_RECORD_STREAMS=1
+ENV NCCL_IB_SL=1
+ENV NCCL_PROTO=LL128
```
Comment (nluehr) [inline .github/container/Dockerfile.jax:54]: Lets drop the NCCL_PROTO option.

Comment (terrykong) [reply .github/container/Dockerfile.jax]: Done

**Comment by nluehr** [comment]
As best I can tell, NCCL_AVOID_RECORD_STREAMS is a feature of pytorch rather than NCCL. (It seems it's now been renamed to TORCH_NCCL_AVOID_RECORD_STREAMS [here](https://github.com/pytorch/pytorch/commit/0781188e64c7065dc7c3a9607ebb4cf0ee0e3d16)).
So I think it's safe to drop it entirely.

**Review by abhinavgoel95 — APPROVED** [review]

**Review by yhtang — CHANGES_REQUESTED** [review]
Did we figure out a source for the undocumented `NCCL_AVOID_RECORD_STREAMS` variable?

**Comment by terrykong** [comment]
@abhinavgoel95

**Comment by abhinavgoel95** [comment]
@yhtang @terrykong it is safe to drop NCCL_AVOID_RECORD_STREAMS. We do not need to upstream this. It is a PyTorch specific feature.

**Comment by terrykong** [comment]
I removed mention of that env var. Can I re-request your review @yhtang ?

**Review by yhtang — APPROVED** [review]

**Pull request merged**